text/0000-colocated-tests.md

```diff
@@ -0,0 +1,191 @@
+This same flexibility should be available for routes, controllers, etc -- but is out of scope for this RFC as other resolver implementations may be in-flight.
+Avoiding implementing `componentStructure` for the rest of the `app` directory does not have any impact on test co-location as the resolver is not involved with finding test file locations. For example, it would be valid to have `app/routes/{route-name}.js` and tests located at `app/routes/__tests__/{route-name}.test.js`.
+
+### Addons, in-repo Addons and Engines
+
+For normal addons, living as their own entire project, the lookup rules for apps will apply.
+For in-repo addons and engines, there is the possibility that the implementation of this feature would enable things not possible today.
+For example, if the "test finder" followed the glob: `**/*(\.test.js|-test.js)`, (excluding node\_modules, etc), that would include tests that are in in-repo addons and engines,
+which today, do not support their own tests -- the in-repo addon or engine still wouldn't support their own tests, exactly, as they would run as part of the parent
+app's test suite. This might be desired behavior as it would allow the in-repo addon and engine to finally co-locate their tests with the behavior that those units implement.
+
+
+### Reference: Where do the existing tests get migrated to?
+While tests in `{app,addon}/` could live anywhere, these are to be the default locations for the generators to use.
+
+#### Route
+
+previously:
+ - `{app,addon}/routes/{route-name}.js`
+ - `tests/unit/{route-name}-test.js`
+now:
+ - `{app,addon}/routes/{route-name}.js`
+ - `{app,addon}/routes/{route-name}.test.js`
+
+#### Service
+
+previously:
+ - `{app,addon}/services/{service-name}.js`
+ - `tests/unit/{service-name}-test.js`
+now:
+ - `{app,addon}/service/{service-name}.js`
+ - `{app,addon}/service/{service-name}.test.js`
+
+#### Controller
+
+previously:
+ - `{app,addon}/controllers/{controller-name}.js`
+ - `tests/unit/{controller-name}-test.js`
+now:
+ - `{app,addon}/controllers/{controller-name}.js`
+ - `{app,addon}/controllers/{controller-name}.test.js`
+
+#### Component
+
+previously:
+ - `{app,addon}/components/{component-name}.js`
+ - `{app,addon}/components/{component-name}.hbs`
+ - `tests/integration/components/{component-name}-test.js`
+now:
+ - `app/components/{component-name}.js`
+ - `app/components/{component-name}.hbs`
+ - `app/components/{component-name}.test.js`
+
+or using `"componentStructure": "nested"`:
+ - `{app,addon}/components/{component-name}/index.js`
+ - `{app,addon}/components/{component-name}/index.hbs`
+ - `{app,addon}/components/{component-name}/index.test.js`
+
+#### Helpers
+
+previously:
+ - `{app,addon}/helpers/{helper-name}.js`
+ - `tests/integration/helpers/{helper-name}-test.js`
+now:
+ - `{app,addon}/helpers/{helper-name}.js`
+ - `{app,addon}/helpers/{helper-name}.test.js`
```

Comment:
what if I want unit _and_ rendering test?

- `{app,addon}/helpers/{helper-name}.unit.test.js`
- `{app,addon}/helpers/{helper-name}.rendering.test.js`

like this? If so, please add as examples

Reply:
do unit tests for helpers exist?

Reply:
Why use different files? Why not

```javascript
module('...', function() {
  module('...', function(hooks) {
    setupTest(hooks);
    // ...
  });
  module('...', function(hooks) {
    setupRenderingTest(hooks);
    // ...
  });
});
```

Reply:
@NullVoxPopuli I believe before RFC232, helpers were tested by importing the helper function as testing it as a Javascript function. I don't think that makes sense now that we have rendering tests, but there are probably still legacy helper unit tests out there.

Reply:
I mean, those legacy tests will still work. because (imo), it may not be the happy path, people could either leave their files in `tests/` or manually rename?

Reply:
@NullVoxPopuli I agree with you, was just trying to answer your direct question. In fact, if mixing unit and rendering tests, I'd either put them in different modules in the same file (see above), or just put them all in a rendering test module -- you can do "unit tests" in a rendering test environment -- just don't call `render()`!

I think the one case where tests don't mix in the same module is application and rendering tests. As rwjblue [likes to point out](https://github.com/emberjs/rfcs/pull/575#pullrequestreview-355235832), `setupTest()` and `setupApplicationTest()` do almost the same thing -- both fully boot an application. `setupRenderingTest()` is the odd one since it sets up a sandboxed rendering environment. And I have on occasion written both application and rendering tests for the same subject, but I'd argue those are pretty rare, and nested modules (or just custom test file naming) gives a clean answer there.

So IMO maaaaaaaybe there's a "how we teach this" point in here, but doesn't seem necessary since anything like this is pretty off the beaten path and also has a fine answer.